OsmAnd/src/net/osmand/plus/mapmarkers/ItineraryDataHelper.java

```diff
@@ -0,0 +1,421 @@
+
+import androidx.annotation.NonNull;
+import androidx.core.content.ContextCompat;
+
+import net.osmand.FileUtils;
+import net.osmand.GPXUtilities;
+import net.osmand.GPXUtilities.GPXExtensionsReader;
+import net.osmand.GPXUtilities.GPXExtensionsWriter;
+import net.osmand.GPXUtilities.GPXFile;
+import net.osmand.GPXUtilities.WptPt;
+import net.osmand.IndexConstants;
+import net.osmand.PlatformUtil;
+import net.osmand.data.LatLon;
+import net.osmand.data.PointDescription;
+import net.osmand.plus.FavouritesDbHelper;
+import net.osmand.plus.OsmandApplication;
+import net.osmand.plus.Version;
+import net.osmand.util.Algorithms;
+
+import org.apache.commons.logging.Log;
+import org.xmlpull.v1.XmlPullParser;
+import org.xmlpull.v1.XmlPullParserException;
+import org.xmlpull.v1.XmlSerializer;
+
+import java.io.File;
+import java.io.IOException;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.TimeZone;
+
+import static net.osmand.GPXUtilities.readText;
+import static net.osmand.GPXUtilities.writeNotNullText;
+import static net.osmand.plus.FavouritesDbHelper.backup;
+import static net.osmand.plus.mapmarkers.MapMarkersDbHelper.DB_NAME;
+import static net.osmand.util.MapUtils.createShortLinkString;
+
+public class ItineraryDataHelper {
+
+	private static final Log log = PlatformUtil.getLog(ItineraryDataHelper.class);
+
+	private static final String VISITED_DATE = "visited_date";
+	private static final String CREATION_DATE = "creation_date";
+
+	private static final String CATEGORIES_SPLIT = ",";
+	private static final String FILE_TO_SAVE = "itinerary.gpx";
+	private static final String FILE_TO_BACKUP = "itinerary_bak.gpx";
+	private static final String ITINERARY_ID = "itinerary_id";
+	private static final String ITINERARY_GROUP = "itinerary_group";
+	private static final String GPX_KEY = "gpx";
+	private static final String FAVOURITES_KEY = "favourites_group";
+
+	private static final SimpleDateFormat GPX_TIME_FORMAT = new SimpleDateFormat(GPXUtilities.GPX_TIME_FORMAT, Locale.US);
+
+	static {
+		GPX_TIME_FORMAT.setTimeZone(TimeZone.getTimeZone("UTC"));
+	}
+
+	private OsmandApplication app;
+	private MapMarkersHelper mapMarkersHelper;
+
+	public ItineraryDataHelper(OsmandApplication app, MapMarkersHelper mapMarkersHelper) {
+		this.app = app;
+		this.mapMarkersHelper = mapMarkersHelper;
+	}
+
+	private File getInternalFile() {
+		return app.getFileStreamPath(FILE_TO_BACKUP);
+	}
+
+	public File getExternalFile() {
+		return new File(app.getAppPath(null), FILE_TO_SAVE);
+	}
+
+	public File getBackupFile() {
+		return FavouritesDbHelper.getBackupFile(app, "itinerary_bak_");
+	}
+
+	public List<MapMarkersGroup> loadGroups() {
+		File internalFile = getInternalFile();
+		if (!internalFile.exists()) {
```

Comment:
should be commented that's only for one release... Probably that code should be moved to InitializeVersion to do 1 migration.
If file is not present, then automatically 1 group should be created only with markers